pulser/simulation/simresults.py

```diff
@@ -89,7 +331,7 @@ def get_final_state(self, reduce_to_basis: Optional[str] = None,
             TypeError: If trying to reduce to a basis that would eliminate
                 states with significant occupation probabilites.
         """
-        final_state = self._states[-1].copy()
+        final_state = cast(qutip.Qobj, self._results[t].copy())
```

Comment:
nit: Don't call it `final_state` anymore, I guess. 

Also, what happens when the user specifies a `t` that is not in `sim_times`?

Reply:
Just an out-of-range exception ; should we add checks everytime a time (more like "step" parameter, since it's an integer) is passed to a method ?

Reply:
You should be consistent and always use time. I would make a method that takes a value of `t` and returns the proper index for `_results`, or error otherwise. I would run `t` through it on `get_state` and `sample_state`